### PR TITLE
Working on stuff about constructors in 5.2

### PR DESCRIPTION
**Pull request by gigamonkey**
This is more by way of a first draft of this new material. I haven't gone over it with a fine-tooth comb yet but wanted to let you see what I've got so far.

### CODE REVIEW

File: _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst [hunk 1]
```diff
@@ -40,76 +40,137 @@ Writing Constructors
...
+   }
 
-     // constructor: initialize all 3 instance variables to parameters
-     public Person(String initName, String initEmail, String initPhone)
-     {
-        name = initName;
-        email = initEmail;
-        phoneNumber = initPhone;
-     }
+Sometimes you still might still even want to provide a no-argument constructor.
```
Comment (bhoffman0) [inline _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst:154]:
```suggestion
Sometimes you might still even want to provide a no-argument constructor.
```

File: _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst [hunk 2]
```diff
@@ -40,76 +40,137 @@ Writing Constructors
...
+If there’s a valid object that you can create without any arguments, you could
+write a no-argument constructor for person like:
```
Comment (bhoffman0) [inline _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst:156]:
```suggestion
write a no-argument constructor for Person like:
```

File: _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst [hunk 3]
```diff
@@ -159,148 +220,163 @@ If there are no constructors written for a class, Java provides a no-argument **
...
-     private int denominator;
+   :language: java
+   :autograde: unittest
+
+   The following class defines a ``Fraction`` with the instance variables
+   ``numerator`` and ``denominator``. It uses 2 constructors. Note that the
+   no-argument constructor sets the default instance variable values to 1 rather
+   than 0 since a fraction with 0 in the denominator is not valid. Try to guess
+   what it will print before you run it. Hint! Remember to start with the
+   ``main`` method! You can also view it in the |Java visualizer| by clicking on
```
Comment (bhoffman0) [inline _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst:231]:
```suggestion
   ``main`` method! You can also view it in the Java visualizer by clicking on
```

File: _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst [hunk 4]
```diff
@@ -159,148 +220,163 @@ If there are no constructors written for a class, Java provides a no-argument **
...0public%20Fraction%28int%20initNumerator,%20int%20initDenominator%29%0A%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20numerator%20%3D%20initNumerator%3B%0A%20%20%20%20%20%20%20%20denominator%20%3D%20initDenominator%3B%0A%20%20%20%20%20%7D%0A%20%20%20%20%20%0A%20%20%20%20%20//%20Print%20fraction%0A%20%20%20%20%20public%20void%20print%28%29%0A%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20System.out.println%28numerator%20%2B%20%22/%22%20%2B%20denominator%29%3B%0A%20%20%20%20%20%7D%0A%20%20%20%20%20%0A%20%20%20%20%20//%20main%20method%20for%20testing%0A%20%20%20%20%20public%20static%20void%20main%28String%5B%5D%20args%29%0A%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20Fraction%20f1%20%3D%20new%20Fraction%28%29%3B%0A%20%20%20%20%20%20%20%20Fraction%20f2%20%3D%20new%20Fraction%281,2%29%3B%0A%20%20%20%20%20%20%20%20//%20What%20will%20these%20print%20out%3F%0A%20%20%20%20%20%20%20%20f1.print%28%29%3B%0A%20%20%20%20%20%20%20%20f2.print%28%29%3B%0A%20%20%20%20%20%7D%0A%20%20%7D&cumulative=false&curInstr=28&heapPrimitives=nevernest&mode=display&origin=opt-frontend.js&py=java&rawInputLstJSON=%5B%5D&textReferences=false&curInstr=0" target="_blank"  style="text-decoration:underline">Java visualizer</a>
```
Comment (bhoffman0) [inline _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst:220]: Could delete this one.

File: _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst [hunk 5]
```diff
@@ -159,148 +220,163 @@ If there are no constructors written for a class, Java provides a no-argument **
...
-            boolean passed = getResults(expect, output, "Running main");
-            assertTrue(passed);
-        }
+   public class RunestoneTests extends CodeTestHelper
+   {
+       @Test
+       public void testMain() throws IOException
+       {
+           String output = getMethodOutput("main");
+           String expect = "Car model: Ford\nCar year: 2019\nCar model: Honda\nCar year: 2010";
```
Comment (bhoffman0) [inline _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst:360]:
```suggestion
           String expect = "Car model: Ford\nCar year: 2023\nCar model: Honda\nCar year: 2010";
```

File: _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst [hunk 6]
```diff
@@ -159,148 +220,163 @@ If there are no constructors written for a class, Java provides a no-argument **
...
+       }
 
-Constructors are used to set the initial state of an object by initializing its instance variables. The examples above have instance variables that are primitive types, but you can have other objects, reference types, as instance variables. For example, a Person class could have an Address object as an instance variable, and the Address class could have String instance variables for the street, city, and state.
+   }
 
-(Advanced AP Topic Warning) When you pass object references as parameters to constructors or methods, they become aliases for the original object and can change it. If a constructor has an object instance variable, it can copy   the referenced object in the parameter using new and the constructor of the referenced object like below so that it does not change the state of the original object. You will see more examples like this in later lessons.
+(Advanced AP Topic Warning) When you pass object references as parameters to
+constructors or methods, those references refer to the same objects as the
+references in the caller. If the objects are immutable, like ``String`` objects
+it doesn’t matter at all. On the other hand, if the objects are **mutable**,
```
Comment (bhoffman0) [inline _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst:371]: This is a great explanation!

**Comment by gigamonkey** [comment]
Okay. I'm done with this, I think.

File: _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst [hunk 7]
```diff
@@ -40,77 +40,138 @@ Writing Constructors
...
-  constructors. I'd much rather explain that you can write a no-arg constructor
-  but only should if you can create a meaningful default object. I'd argue that
-  a Person with no name, no email, and no phone number is not a meaningful
-  object. What do you think? -Peter
+The job of a constructor is to set the initial values for the object’s instance
+variables to useful values. But what does “useful” mean? Sometimes we describe
+the values of all an object's instance variables at a given time as the object's
+**state**. And we say an object is in a **valid state** when all its instance
+variables have values that let us use the object by invoking its public methods.
+So another way to describe the job of a constructor is to set the the object’s
```
Comment (bhoffman0) [inline _sources/Unit5-Writing-Classes/topic-5-2-writing-constructors.rst:94]:
```suggestion
So another way to describe the job of a constructor is to set the object’s
```

**Review by bhoffman0 — APPROVED** [review]
Looks good!

**Pull request merged**